Serialization error for "Map<K, V[]>" (Regression Fury 0.10.3 -> Fory 0.11.0)

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

The following class was serialized fine with **Fury 0.10.3**, but fails with a compile error with **Fory 0.11.0**.

Seems to be a regression caused by one of the changes.

```java
    public class State<K extends Comparable<K>, V> {
        Map<K, V[]> map;
        public State(Map<K, V[]> map) {
            this.map = map;
        }
    }
```

The generated code is wrong and fails when compiled:
`/* 0027 */   private final org.apache.fory.type.GenericType object[]1;`


### Component(s)

Java

### Minimal reproduce step

```java
import org.apache.fory.Fory;
import org.apache.fory.config.Language;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Map;

public class SerTest {

    public static void main(String[] args) throws IOException {

        Fory fory = Fory.builder()
                .withLanguage(Language.JAVA)
                .requireClassRegistration(false)
                .build();

        State original = new State(Map.of("foo", new String[]{"bar"}));
        System.out.println("ORIGINAL: " + original);

        try (var out = new ByteArrayOutputStream()) {
            fory.serialize(out, original);
            byte[] data = out.toByteArray();
            State copy = (State) fory.deserialize(data);
            System.out.println("COPY: " + copy);
        }

    }

    public static class State<K extends Comparable<K>, V> {

        Map<K, V[]> map;

        public State(Map<K, V[]> map) {
            this.map = map;
        }

    }

}
```


### What did you expect to see?

Should run without errors as it was the case in Fury 0.10.3

### What did you see instead?

```
Exception in thread "main" java.lang.RuntimeException: Create sequential serializer failed, 
class: class SerTest$State
	at org.apache.fory.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:52)
	at org.apache.fory.resolver.ClassResolver.lambda$getObjectSerializerClass$2(ClassResolver.java:1126)
	at org.apache.fory.builder.JITContext.registerSerializerJITCallback(JITContext.java:121)
	at org.apache.fory.resolver.ClassResolver.getObjectSerializerClass(ClassResolver.java:1124)
	at org.apache.fory.resolver.ClassResolver.getSerializerClass(ClassResolver.java:1039)
	at org.apache.fory.resolver.ClassResolver.getSerializerClass(ClassResolver.java:917)
	at org.apache.fory.resolver.ClassResolver.createSerializer(ClassResolver.java:1390)
	at org.apache.fory.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1312)
	at org.apache.fory.Fory.write(Fory.java:401)
	at org.apache.fory.Fory.serialize(Fory.java:325)
	at org.apache.fory.Fory.lambda$serialize$0(Fory.java:340)
	at org.apache.fory.Fory.serializeToStream(Fory.java:1528)
	at org.apache.fory.Fory.serialize(Fory.java:340)
	at SerTest.main(SerTest.java:21)
Caused by: org.apache.fory.codegen.CodegenException: Compile error: 
.SerTest_StateForyCodec_0:
/* 0001 */ import org.apache.fory.Fory;
/* 0002 */ import org.apache.fory.memory.MemoryBuffer;
/* 0003 */ import org.apache.fory.resolver.NoRefResolver;
/* 0004 */ import org.apache.fory.memory.Platform;
/* 0005 */ import org.apache.fory.resolver.ClassInfo;
/* 0006 */ import org.apache.fory.resolver.ClassInfoHolder;
/* 0007 */ import org.apache.fory.resolver.ClassResolver;
/* 0008 */ import org.apache.fory.builder.Generated;
/* 0009 */ import org.apache.fory.serializer.CodegenSerializer.LazyInitBeanSerializer;
/* 0010 */ import org.apache.fory.serializer.EnumSerializer;
/* 0011 */ import org.apache.fory.serializer.Serializer;
/* 0012 */ import org.apache.fory.serializer.StringSerializer;
/* 0013 */ import org.apache.fory.serializer.ObjectSerializer;
/* 0014 */ import org.apache.fory.serializer.CompatibleSerializer;
/* 0015 */ import org.apache.fory.serializer.collection.AbstractCollectionSerializer;
/* 0016 */ import org.apache.fory.serializer.collection.AbstractMapSerializer;
/* 0017 */ import org.apache.fory.builder.Generated.GeneratedObjectSerializer;
/* 0018 */ 
/* 0019 */ public final class SerTest_StateForyCodec_0 extends GeneratedObjectSerializer {
/* 0020 */ 
/* 0021 */   private final NoRefResolver refResolver;
/* 0022 */   private final ClassResolver classResolver;
/* 0023 */   private final StringSerializer strSerializer;
/* 0024 */   private Fory fory;
/* 0025 */   private ClassInfo mapClassInfo;
/* 0026 */   private final org.apache.fory.type.GenericType comparable0;
/* 0027 */   private final org.apache.fory.type.GenericType object[]1;
/* 0028 */   private ClassInfo comparableClassInfo;
/* 0029 */   private ClassInfo arrClassInfo;
/* 0030 */   private final ClassInfoHolder map3ClassInfoHolder;
/* 0031 */   private final ClassInfoHolder comparable1ClassInfoHolder;
/* 0032 */   private final org.apache.fory.serializer.ArraySerializers.ObjectArraySerializer objectArraySerializer;
/* 0033 */   private final ClassInfoHolder arr1ClassInfoHolder;
/* 0034 */ 
/* 0035 */   public SerTest_StateForyCodec_0(Fory fory, Class classType) {
/* 0036 */       super(fory, classType);
/* 0037 */       this.fory = fory;
/* 0038 */       fory.getClassResolver().setSerializerIfAbsent(classType, this);
/* 0039 */   
/* 0040 */       org.apache.fory.resolver.RefResolver refResolver0 = fory.getRefResolver();
/* 0041 */       refResolver = ((NoRefResolver)refResolver0);
/* 0042 */       classResolver = fory.getClassResolver();
/* 0043 */       strSerializer = fory.getStringSerializer();
/* 0044 */       mapClassInfo = classResolver.nilClassInfo();
/* 0045 */       org.apache.fory.type.GenericType genericType = classResolver.getGenericTypeInStruct(super.type, "K");
/* 0046 */       comparable0 = genericType;
/* 0047 */       org.apache.fory.type.GenericType genericType1 = classResolver.getGenericTypeInStruct(super.type, "V[]");
/* 0048 */       object[]1 = genericType1;
/* 0049 */       comparableClassInfo = classResolver.nilClassInfo();
/* 0050 */       arrClassInfo = classResolver.nilClassInfo();
/* 0051 */       map3ClassInfoHolder = classResolver.nilClassInfoHolder();
/* 0052 */       comparable1ClassInfoHolder = classResolver.nilClassInfoHolder();
/* 0053 */       objectArraySerializer = ((org.apache.fory.serializer.ArraySerializers.ObjectArraySerializer)classResolver.getRawSerializer(java.lang.Object[].class));
/* 0054 */       arr1ClassInfoHolder = classResolver.nilClassInfoHolder();
/* 0055 */   }
/* 0056 */ 
/* 0057 */   private AbstractMapSerializer writeMapClassInfo(MemoryBuffer memoryBuffer, java.util.Map map1) {
/* 0058 */       ClassResolver classResolver = this.classResolver;
/* 0059 */       Class value = mapClassInfo.getCls();
/* 0060 */       Class cls = map1.getClass();
/* 0061 */       if ((value != cls)) {
/* 0062 */           mapClassInfo = classResolver.getClassInfo(cls);
/* 0063 */       }
/* 0064 */       classResolver.writeClassInfo(memoryBuffer, mapClassInfo);
/* 0065 */       return ((AbstractMapSerializer)mapClassInfo.getSerializer());
/* 0066 */   }
/* 0067 */ 
/* 0068 */   private java.util.Map.Entry writeChunk(MemoryBuffer memoryBuffer1, java.util.Map.Entry entry, java.util.Iterator iterator) {
/* 0069 */       ClassResolver classResolver = this.classResolver;
/* 0070 */       Comparable key = (Comparable)entry.getKey();
/* 0071 */       java.lang.Object[] value0 = (java.lang.Object[])entry.getValue();
/* 0072 */       Class keyType = key.getClass();
/* 0073 */       Class valueType = value0.getClass();
/* 0074 */       memoryBuffer1.writeInt16(((short)-1));
/* 0075 */       int chunkSizeOffset = memoryBuffer1.writerIndex() - 1;
/* 0076 */       int chunkHeader = 0;
/* 0077 */       if ((comparableClassInfo.getCls() != keyType)) {
/* 0078 */           comparableClassInfo = classResolver.getClassInfo(keyType);
/* 0079 */       }
/* 0080 */       classResolver.writeClassInfo(memoryBuffer1, comparableClassInfo);
/* 0081 */       Serializer serializer = comparableClassInfo.getSerializer();
/* 0082 */       if ((arrClassInfo.getCls() != valueType)) {
/* 0083 */           arrClassInfo = classResolver.getClassInfo(valueType);
/* 0084 */       }
/* 0085 */       classResolver.writeClassInfo(memoryBuffer1, arrClassInfo);
/* 0086 */       Serializer serializer1 = arrClassInfo.getSerializer();
/* 0087 */       memoryBuffer1.putByte((chunkSizeOffset - 1), chunkHeader);
/* 0088 */       int chunkSize = 0;
/* 0089 */       while (true) {
/* 0090 */         if (((((key == null) || (value0 == null)) || (key.getClass() != keyType)) || (value0.getClass() != valueType))) {
/* 0091 */             break;
/* 0092 */         }
/* 0093 */         serializer.write(memoryBuffer1, key);
/* 0094 */         serializer1.write(memoryBuffer1, value0);
/* 0095 */         chunkSize = (chunkSize + 1);
/* 0096 */         if (iterator.hasNext()) {
/* 0097 */             entry = ((java.util.Map.Entry)iterator.next());
/* 0098 */             key = ((Comparable)entry.getKey());
/* 0099 */             value0 = ((java.lang.Object[])entry.getValue());
/* 0100 */         } else {
/* 0101 */             entry = null;
/* 0102 */             break;
/* 0103 */         }
/* 0104 */         if ((chunkSize == 255)) {
/* 0105 */             break;
/* 0106 */         }
/* 0107 */       }
/* 0108 */       memoryBuffer1.putByte(chunkSizeOffset, chunkSize);
/* 0109 */       return entry;
/* 0110 */   }
/* 0111 */ 
/* 0112 */   private void writeFields(MemoryBuffer memoryBuffer2, SerTest.State state1) {
/* 0113 */       java.util.Map map00 = SerTest_StateForyAccessor_0.map(state1);
/* 0114 */       if ((map00 == null)) {
/* 0115 */           memoryBuffer2.writeByte(((byte)-3));
/* 0116 */       } else {
/* 0117 */           memoryBuffer2.writeByte(((byte)-1));
/* 0118 */           AbstractMapSerializer abstractMapSerializer = this.writeMapClassInfo(memoryBuffer2, map00);
/* 0119 */           if (abstractMapSerializer.supportCodegenHook()) {
/* 0120 */               java.util.Map map2 = abstractMapSerializer.onMapWrite(memoryBuffer2, map00);
/* 0121 */               if ((!map2.isEmpty())) {
/* 0122 */                   java.util.Iterator iterator0 = map2.entrySet().iterator();
/* 0123 */                   java.util.Map.Entry entry0 = (java.util.Map.Entry)iterator0.next();
/* 0124 */                   while ((entry0 != null)) {
/* 0125 */                     entry0 = abstractMapSerializer.writeJavaNullChunkGeneric(memoryBuffer2, entry0, iterator0, comparable0, object[]1);
/* 0126 */                     if ((entry0 != null)) {
/* 0127 */                         entry0 = this.writeChunk(memoryBuffer2, entry0, iterator0);
/* 0128 */                     }
/* 0129 */                   }
/* 0130 */               }
/* 0131 */           } else {
/* 0132 */               abstractMapSerializer.write(memoryBuffer2, map00);
/* 0133 */           }
/* 0134 */       }
/* 0135 */   }
/* 0136 */ 
/* 0137 */   private int readChunk(MemoryBuffer memoryBuffer3, int value1, int value2, java.util.Map map4) {
/* 0138 */       ClassResolver classResolver = this.classResolver;
/* 0139 */       int chunkSize0 = memoryBuffer3.readUnsignedByte();
/* 0140 */       Serializer serializer0 = classResolver.readClassInfo(memoryBuffer3, comparable1ClassInfoHolder).getSerializer();
/* 0141 */       Serializer serializer2;
/* 0142 */       if (((value2 & 32) != 0)) {
/* 0143 */           serializer2 = objectArraySerializer;
/* 0144 */       } else {
/* 0145 */           serializer2 = classResolver.readClassInfo(memoryBuffer3, arr1ClassInfoHolder).getSerializer();
/* 0146 */       }
/* 0147 */       for (int i = 0; i < chunkSize0; i+=1) {
/* 0148 */         Object object = serializer0.read(memoryBuffer3);
/* 0149 */         Object object1 = serializer2.read(memoryBuffer3);
/* 0150 */         map4.put(object, object1);
/* 0151 */         value1 = (value1 - 1);
/* 0152 */       }
/* 0153 */       if ((value1 > 0)) {
/* 0154 */           return ((value1 << 8) | memoryBuffer3.readUnsignedByte());
/* 0155 */       } else {
/* 0156 */           return 0;
/* 0157 */       }
/* 0158 */   }
/* 0159 */ 
/* 0160 */   private void readFields(SerTest.State state2, MemoryBuffer memoryBuffer4) {
/* 0161 */       if ((memoryBuffer4.readByte() != ((byte)-3))) {
/* 0162 */           AbstractMapSerializer mapSerializer = (AbstractMapSerializer)classResolver.readClassInfo(memoryBuffer4, map3ClassInfoHolder).getSerializer();
/* 0163 */           Object object2;
/* 0164 */           if (mapSerializer.supportCodegenHook()) {
/* 0165 */               java.util.Map map5 = mapSerializer.newMap(memoryBuffer4);
/* 0166 */               int size = mapSerializer.getAndClearNumElements();
/* 0167 */               int value3;
/* 0168 */               if ((size == 0)) {
/* 0169 */                   value3 = 0;
/* 0170 */               } else {
/* 0171 */                   value3 = memoryBuffer4.readUnsignedByte();
/* 0172 */               }
/* 0173 */               while ((size > 0)) {
/* 0174 */                 long sizeAndHeader = mapSerializer.readJavaNullChunk(memoryBuffer4, map5, value3, size, null, null);
/* 0175 */                 value3 = ((int)(sizeAndHeader & 255));
/* 0176 */                 size = ((int)(sizeAndHeader >>> 8));
/* 0177 */                 if ((size == 0)) {
/* 0178 */                     break;
/* 0179 */                 }
/* 0180 */                 int value4 = this.readChunk(memoryBuffer4, size, value3, map5);
/* 0181 */                 value3 = (value4 & 255);
/* 0182 */                 size = (value4 >>> 8);
/* 0183 */               }
/* 0184 */               object2 = mapSerializer.onMapRead(map5);
/* 0185 */           } else {
/* 0186 */               Object object0 = mapSerializer.read(memoryBuffer4);
/* 0187 */               object2 = object0;
/* 0188 */           }
/* 0189 */           
/* 0190 */           SerTest_StateForyAccessor_0.map(state2, ((java.util.Map)object2));
/* 0191 */       } else {
/* 0192 */           SerTest_StateForyAccessor_0.map(state2, null);
/* 0193 */       }
/* 0194 */   }
/* 0195 */ 
/* 0196 */   @Override public final void write(MemoryBuffer buffer, Object obj) {
/* 0197 */       SerTest.State state3 = (SerTest.State)obj;
/* 0198 */       this.writeFields(buffer, state3);
/* 0199 */   }
/* 0200 */ 
/* 0201 */   @Override public final Object read(MemoryBuffer buffer) {
/* 0202 */       Object instance = Platform.newInstance(SerTest$State.class);
/* 0203 */       SerTest.State state4 = (SerTest.State)instance;
/* 0204 */       refResolver.reference(state4);
/* 0205 */       this.readFields(state4, buffer);
/* 0206 */       return state4;
/* 0207 */   }
/* 0208 */ 
/* 0209 */ }
	at org.apache.fory.codegen.JaninoUtils.toBytecode(JaninoUtils.java:133)
	at org.apache.fory.codegen.JaninoUtils.toBytecode(JaninoUtils.java:73)
	at org.apache.fory.codegen.CodeGenerator.compile(CodeGenerator.java:145)
	at org.apache.fory.builder.CodecUtils.loadOrGenCodecClass(CodecUtils.java:89)
	at org.apache.fory.builder.CodecUtils.loadOrGenObjectCodecClass(CodecUtils.java:44)
	at org.apache.fory.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:49)
	... 13 more
Caused by: org.apache.fory.shaded.org.codehaus.commons.compiler.CompileException: File '/SerTest_StateForyCodec_0.java', Line 27, Column 58: ';' expected instead of '1'
	at org.apache.fory.shaded.org.codehaus.janino.TokenStreamImpl.compileException(TokenStreamImpl.java:362)
	at org.apache.fory.shaded.org.codehaus.janino.TokenStreamImpl.read(TokenStreamImpl.java:149)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.read(Parser.java:3804)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.parseClassBodyDeclaration(Parser.java:1070)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.parseClassBody(Parser.java:857)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.parseClassDeclarationRest(Parser.java:747)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.parsePackageMemberTypeDeclarationRest(Parser.java:493)
	at org.apache.fory.shaded.org.codehaus.janino.Parser.parseAbstractCompilationUnit(Parser.java:268)
	at org.apache.fory.shaded.org.codehaus.janino.Compiler.parseAbstractCompilationUnit(Compiler.java:316)
	at org.apache.fory.shaded.org.codehaus.janino.Compiler.compile2(Compiler.java:236)
	at org.apache.fory.shaded.org.codehaus.janino.Compiler.compile(Compiler.java:213)
	at org.apache.fory.codegen.JaninoUtils.toBytecode(JaninoUtils.java:115)
	... 18 more
```


### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fory-core/src/main/java/org/apache/fory/builder/BaseObjectCodecBuilder.java
+++ b/java/fory-core/src/main/java/org/apache/fory/builder/BaseObjectCodecBuilder.java
@@ -666,9 +666,17 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
     String name =
         namesForSharedGenericTypeFields.computeIfAbsent(
             typeRef,
-            k ->
-                StringUtils.uncapitalize(typeRef.getRawType().getSimpleName())
-                    + namesForSharedGenericTypeFields.size());
+            k -> {
+              String prefix;
+              if (typeRef.getRawType().isArray()) {
+                Tuple2<Class<?>, Integer> info =
+                    TypeUtils.getArrayComponentInfo(typeRef.getRawType());
+                prefix = info.f0.getSimpleName() + "_arr" + info.f1;
+              } else {
+                prefix = typeRef.getRawType().getSimpleName();
+              }
+              return StringUtils.uncapitalize(prefix) + namesForSharedGenericTypeFields.size();
+            });
 
     return getOrCreateField(
         false,

--- a/java/fory-core/src/test/java/org/apache/fory/serializer/collection/MapSerializersTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/serializer/collection/MapSerializersTest.java
@@ -1028,4 +1028,24 @@ public class MapSerializersTest extends ForyTestBase {
     Object object = fory2.deserialize(bytes);
     assertEquals(object, o);
   }
+
+  @Data
+  @AllArgsConstructor
+  public static class State<K extends Comparable<K>, V> {
+    Map<K, V[]> map;
+  }
+
+  @Test
+  public void testChunkArrayGeneric() {
+    Fory fory = Fory.builder().withLanguage(Language.JAVA).requireClassRegistration(false).build();
+    State original = new State(ofHashMap("foo", new String[] {"bar"}));
+    State state = serDe(fory, original);
+    Assert.assertEquals(state.map.get("foo"), new String[] {"bar"});
+
+    State state1 = new State(ofHashMap("foo", null, "bar", new String[] {"bar"}));
+    byte[] bytes = fory.serialize(state1);
+    Fory fory2 = builder().withCodegen(false).build();
+    State state2 = (State) fory2.deserialize(bytes);
+    Assert.assertEquals(state2.map.get("bar"), new String[] {"bar"});
+  }
 }

